Errors while building OpenCV 3.0.0 with -D BUILD_TBB flag

Transferred from http://code.opencv.org/issues/4438
```
|| Dhruv Vyas on 2015-06-26 09:14
|| Priority: Normal
|| Affected: branch 'master' (3.0-dev)
|| Category: None
|| Tracker: Bug
|| Difficulty: 
|| PR: 
|| Platform: Any / Linux
```

Errors while building OpenCV 3.0.0 with -D BUILD_TBB flag
-----------
```
While compiling OpenCV 3.0 with -D BUILD_TBB=ON -D WITH_TBB=ON cmake flags on a debian machine. It throws following errors.

[ 56%] Building CXX object modules/core/CMakeFiles/opencv_perf_core.dir/perf/perf_abs.cpp.o [ 58%] Built target opencv_features2d [ 58%] Building CXX object modules/core/CMakeFiles/opencv_perf_core.dir/perf/perf_main.cpp.o ../../lib/libopencv_imgproc.so.3.0.0: undefined reference to tbb::task_group_context::cancel_group_execution()' ../../lib/libopencv_imgproc.so.3.0.0: undefined reference totbb::interface5::internal::task_base::destroy(tbb::task&)' ../../lib/libopencv_core.so.3.0.0: undefined reference to tbb::task_scheduler_init::initialize(int)' ../../lib/libopencv_core.so.3.0.0: undefined reference totbb::task::note_affinity(unsigned short)' ../../lib/libopencv_imgproc.so.3.0.0: undefined reference to tbb::internal::handle_perror(int, char const*)' ../../lib/libopencv_core.so.3.0.0: undefined reference totbb::internal::allocate_child_proxy::allocate(unsigned int) const' ../../lib/libopencv_imgproc.so.3.0.0: undefined reference to tbb::internal::NFS_Free(void*)' ../../lib/libopencv_core.so.3.0.0: undefined reference totbb::task_group_context::is_group_execution_cancelled() const' ../../lib/libopencv_imgproc.so.3.0.0: undefined reference to tbb::internal::throw_exception_v4(tbb::internal::exception_id)' ../../lib/libopencv_core.so.3.0.0: undefined reference totbb::task_scheduler_init::terminate()' ../../lib/libopencv_core.so.3.0.0: undefined reference to tbb::task_scheduler_init::initialize(int, unsigned int)' ../../lib/libopencv_core.so.3.0.0: undefined reference totypeinfo for tbb::task' ../../lib/libopencv_imgproc.so.3.0.0: undefined reference to tbb::internal::allocate_additional_child_of_proxy::allocate(unsigned int) const' ../../lib/libopencv_imgproc.so.3.0.0: undefined reference totbb::task_group_context::reset()' ../../lib/libopencv_core.so.3.0.0: undefined reference to tbb::internal::allocate_continuation_proxy::allocate(unsigned int) const' ../../lib/libopencv_core.so.3.0.0: undefined reference tovtable for tbb::task' ../../lib/libopencv_core.so.3.0.0: undefined reference to tbb::internal::allocate_root_with_context_proxy::free(tbb::task&) const' ../../lib/libopencv_core.so.3.0.0: undefined reference totbb::task_group_context::init()' ../../lib/libopencv_imgproc.so.3.0.0: undefined reference to tbb::internal::NFS_Allocate(unsigned int, unsigned int, void*)' ../../lib/libopencv_core.so.3.0.0: undefined reference totbb::task_scheduler_init::default_num_threads()' ../../lib/libopencv_core.so.3.0.0: undefined reference to tbb::internal::get_initial_auto_partitioner_divisor()' ../../lib/libopencv_core.so.3.0.0: undefined reference totbb::task_group_context::~task_group_context()' ../../lib/libopencv_core.so.3.0.0: undefined reference to `tbb::internal::allocate_root_with_context_proxy::allocate(unsigned int) const' collect2: error: ld returned 1 exit status
```

History
-------
##### Alexander Alekhin on 2015-07-13 11:53
```
Confirm this bug on Ubuntu 12.04 / Fedora 21
-   Status changed from New to Open
```

--- a/3rdparty/tbb/CMakeLists.txt
+++ b/3rdparty/tbb/CMakeLists.txt
@@ -212,7 +212,13 @@ else()
 endif()
 
 ocv_warnings_disable(CMAKE_CXX_FLAGS -Wundef -Wmissing-declarations)
-string(REPLACE "-Werror=non-virtual-dtor" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
+
+# filter out flags that are not handled well by the TBB code
+foreach(var CMAKE_CXX_FLAGS CMAKE_CXX_FLAGS_RELEASE CMAKE_CXX_FLAGS_DEBUG)
+  string(REPLACE "-Werror=non-virtual-dtor" "" ${var} "${${var}}")
+  string(REPLACE "-fvisibility=hidden" "" ${var} "${${var}}")
+  string(REPLACE "-fvisibility-inlines-hidden" "" ${var} "${${var}}")
+endforeach()
 
 if (WIN32)
   set(tbb_debug_postfix "_debug") # to fit pragmas in _windef.h inside TBB

